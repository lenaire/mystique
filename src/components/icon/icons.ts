export { ReactComponent as alert } from "../../../public/alert-icon.svg";
export { ReactComponent as check } from "../../../public/check-icon.svg";
export { ReactComponent as error } from "../../../public/error-icon.svg";
export { ReactComponent as chevron } from "../../../public/chevron-icon.svg";
export { ReactComponent as arrow } from "../../../public/arrow-icon.svg";
export { ReactComponent as close } from "../../../public/close-icon.svg";
