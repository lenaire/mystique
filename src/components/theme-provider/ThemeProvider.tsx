import { FC } from "react";
import {
  DefaultTheme, // This is overridden in types/theme.d.ts
  ThemeProvider as StyledComponentsThemeProvider,
  ThemeProviderProps as StyledComponentsThemeProviderProps,
} from "styled-components";

import { defaultTheme } from "$cmp/theme-provider/theme";
import { ResetGlobalCSS } from "$cmp/theme-provider/ResetGlobalCSS";

export type ThemeProviderProps = Partial<StyledComponentsThemeProviderProps<DefaultTheme>>;

export const ThemeProvider: FC<ThemeProviderProps> = ({ children, theme = defaultTheme }) => {
  return (
    <StyledComponentsThemeProvider theme={theme}>
      <ResetGlobalCSS />
      {children}
    </StyledComponentsThemeProvider>
  );
};
