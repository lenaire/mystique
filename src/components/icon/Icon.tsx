import { forwardRef } from "react";
import styled, { AnyStyledComponent } from "styled-components";
import { CmpProps } from "$types";
import { Direction } from "$utility/enums";
import * as icons from "./icons";
import { alert as Alert } from "./icons";

export type IconType = keyof typeof icons;

const defaultRotationMap = {
  [Direction.Up]: 0,
  [Direction.Down]: 180,
  [Direction.Left]: -90,
  [Direction.Right]: 90,
};

const customRotations: Partial<Record<IconType, RotationMap>> = {
  chevron: {
    [Direction.Up]: 180,
    [Direction.Down]: 0,
    [Direction.Left]: 90,
    [Direction.Right]: -90,
  },
  arrow: {
    [Direction.Up]: 90,
    [Direction.Down]: -90,
    [Direction.Left]: 0,
    [Direction.Right]: 180,
  },
};

type RotationMap = {
  [Direction.Up]: number;
  [Direction.Down]: number;
  [Direction.Left]: number;
  [Direction.Right]: number;
};

const styledIcons = Object.keys(icons).reduce((acc, name) => {
  const Icon = icons[name as IconType] || (Alert as AnyStyledComponent);

  const StyledIcon = styled(Icon)(({ theme }) => ({
    display: "block",
    height: "24px",
    width: "24px",
    fill: theme.palette.common.black,
    flexShrink: 0,
    transition: "transform 0.3s",
  }));

  acc[name as IconType] = StyledIcon;

  return acc;
}, {} as typeof icons);

export type IconProps = CmpProps<
  {
    icon: IconType;
    direction?: Direction;
  } & Omit<React.SVGProps<SVGSVGElement>, "ref">
>;

export const Icon = forwardRef<SVGSVGElement, IconProps>(({ icon, direction, ...rest }, ref) => {
  const title = `A(n) ${icon} icon`;
  const rotationMap = customRotations[icon] || defaultRotationMap;
  const IconCmp = styledIcons[icon];

  return (
    <IconCmp
      title={title}
      {...rest}
      transform={direction ? `rotate(${rotationMap[direction]})` : undefined}
      ref={ref}
    />
  );
});
