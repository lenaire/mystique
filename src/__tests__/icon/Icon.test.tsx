import renderer from "react-test-renderer";
import { ThemeProvider } from "$cmp/theme-provider/ThemeProvider";
import { Icon, IconProps } from "$cmp/icon/Icon";

describe("<Icon />", () => {
  it("renders correctly", () => {
    const props: IconProps = {
      icon: "alert",
    };

    const tree = renderer
      .create(
        <ThemeProvider>
          <Icon {...props} />
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
