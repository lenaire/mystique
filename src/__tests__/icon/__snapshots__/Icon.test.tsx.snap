// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> renders correctly 1`] = `
<div
  className="sc-bcXHqe jkKffL"
  title="A(n) alert icon"
/>
`;
